deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/transferlearning/TransferLearningHelper.java

```diff
@@ -0,0 +1,232 @@
+                            }
+                        }
+                    }
+                }
+            }
+            Layer currentLayer = currentVertex.getLayer();
+            if (currentLayer instanceof FrozenLayer) {
+                //a frozen layer is encountered - should be removed (along with it's inputs)
+                //The question is does it need to be an input to the new smaller unfrozen model or not?
+                if (!seenAsParents.contains(currentName)) {
+                    //not a parent of vertices already seen so needs to be added to the set of inputs
+                    frozenInputVertices.add(currentName);
+                }
+                seenAsParents.add(currentName);
+                VertexIndices[] parentVertices = currentVertex.getInputVertices();
+                //add parents of current frozen vertex to list of seen parents
+                for(int j=0; j<parentVertices.length; j++ ) {
+                    int parentVertexIndex = parentVertices[j].getVertexIndex();
+                    String parentName = origGraph.getVertices()[parentVertexIndex].getVertexName();
+                    seenAsParents.add(parentName);
+                }
+            }
+        }
+
+        TransferLearning.GraphBuilder builder = new TransferLearning.GraphBuilder(origGraph);
+        for (String toRemove: seenAsParents) {
+            if (frozenInputVertices.contains(toRemove)) {
+                builder.removeVertexKeepConnections(toRemove);
+            }
+            else {
+                builder.removeVertexAndConnections(toRemove);
+            }
+        }
+
+        Set<String> frozenInputVerticesSorted = new HashSet<>();
+        frozenInputVerticesSorted.addAll(origGraph.getConfiguration().getNetworkInputs());
+        frozenInputVerticesSorted.removeAll(seenAsParents);
+        //remove input vertices - just to add back in a predictable order
+        for (String existingInput: frozenInputVerticesSorted) {
+            builder.removeVertexKeepConnections(existingInput);
+        }
+        frozenInputVerticesSorted.addAll(frozenInputVertices);
+        //to have a predictable order
+        graphInputs = new ArrayList(frozenInputVerticesSorted);
+        Collections.sort(graphInputs);
+        for (String asInput: frozenInputVerticesSorted) {
+            builder.addInputs(asInput);
+        }
+        unFrozenSubsetGraph = builder.build();
+        copyParamsToGraph();
+
+        if (frozenInputVertices.isEmpty()) {
+            throw new IllegalArgumentException("No frozen layers found");
+        }
+
+    }
+
+    /*
+    private void initHelperMLN() {
+
+        //make smaller graph - loop back in topographical order
+        //find a non frozen vertex that has a frozen parent which will always be a layer vertex
+
+        //outputs are the same
+        //remove some input vertices
+        //set new inputs
+
+    }
+    */
+
+    public MultiDataSet featurize(MultiDataSet input) {
```

Comment:
Even stub javadoc comments would be good here so it picks it up when we generate it.